File: day3/part2/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LINE_LENGTH 1024
#define NUM_DIGITS 12

long long max_joltage(const char *line, int len) {
  if (len < NUM_DIGITS) {
    return 0;
  }

  char result[NUM_DIGITS + 1];
  int result_idx = 0;
  int start = 0;

  for (int i = 0; i < NUM_DIGITS; i++) {

    int max_pos = len - (NUM_DIGITS - i);

    int best_pos = start;
    char best_digit = line[start];

    for (int p = start; p <= max_pos; p++) {
      if (line[p] > best_digit) {
        best_digit = line[p];
        best_pos = p;
      }
    }

    result[result_idx++] = best_digit;
    start = best_pos + 1;
  }

  result[result_idx] = '\0';

  return atoll(result);
}

int main(void) {
  FILE *file = fopen("list.txt", "r");
  if (!file) {
    perror("Failed to open list.txt");
    return 1;
  }

  char line[MAX_LINE_LENGTH];
  long long total = 0;

  while (fgets(line, sizeof(line), file)) {

    int len = strlen(line);
    while (len > 0 && (line[len - 1] == '\n' || line[len - 1] == '\r')) {
      line[--len] = '\0';
    }

    if (len >= NUM_DIGITS) {
      long long joltage = max_joltage(line, len);
      total += joltage;
    }
  }

  fclose(file);

  printf("Total output joltage: %lld\n", total);

  return 0;
}

File: day3/part2/go/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

const (
	MaxLineLength = 1024
	NumDigits     = 12
)

func maxJoltage(line string) int64 {
	if len(line) < NumDigits {
		return 0
	}

	result := make([]byte, 0, NumDigits)
	start := 0

	for i := 0; i < NumDigits; i++ {
		maxPos := len(line) - (NumDigits - i)

		bestPos := start
		bestDigit := line[start]

		for p := start; p <= maxPos; p++ {
			if line[p] > bestDigit {
				bestDigit = line[p]
				bestPos = p
			}
		}

		result = append(result, bestDigit)
		start = bestPos + 1
	}

	joltage, err := strconv.ParseInt(string(result), 10, 64)
	if err != nil {
		return 0
	}
	return joltage
}

func main() {
	file, err := os.Open("../list.txt")
	if err != nil {
		log.Fatalf("Failed to open list.txt: %v", err)
	}
	defer file.Close()

	scanner := bufio.NewScanner(file)
	total := int64(0)

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())

		if len(line) >= NumDigits {
			joltage := maxJoltage(line)
			total += joltage
		}
	}

	if err := scanner.Err(); err != nil {
		log.Fatalf("Scanner error: %v", err)
	}

	fmt.Printf("Total output joltage: %d\n", total)
}
